src/Microsoft.DotNet.Wpf/src/PresentationCore/MS/internal/Text/TextInterface/FontCollectionLoader.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using MS.Internal.Text.TextInterface.Interfaces;
+
+namespace MS.Internal.Text.TextInterface
+{
+    [ClassInterface(ClassInterfaceType.None)]
+    [ComVisible(true)]
+    internal unsafe class FontCollectionLoader : IDWriteFontCollectionLoaderMirror
+    {
+        private const int S_OK = unchecked((int)0L);
+        private const int E_INVALIDARG = unchecked((int)0x80070057L);
+
+        private readonly IFontSourceCollectionFactory _fontSourceCollectionFactory;
+        private readonly FontFileLoader _fontFileLoader;
+
+        public FontCollectionLoader()
+        {
+            Debug.Fail("Assertion failed");
+        }
+
+        public FontCollectionLoader(IFontSourceCollectionFactory fontSourceCollectionFactory, FontFileLoader fontFileLoader)
+        {
+            _fontSourceCollectionFactory = fontSourceCollectionFactory;
+            _fontFileLoader = fontFileLoader;
+        }
+
+        /// <summary>
+        /// Creates a font file enumerator object that encapsulates a collection of font files.
+        /// The font system calls back to this interface to create a font collection.
+        /// </summary>
+        /// <param name="collectionKey">Font collection key that uniquely identifies the collection of font files within
+        /// the scope of the font collection loader being used.</param>
+        /// <param name="collectionKeySize">Size of the font collection key in bytes.</param>
+        /// <param name="fontFileEnumerator">Pointer to the newly created font file enumerator.</param>
+        /// <returns>
+        /// Standard HRESULT error code.
+        /// </returns>
+        [ComVisible(true)]
+        public int CreateEnumeratorFromKey(IntPtr factory, [In] void* collectionKey, [In, MarshalAs(UnmanagedType.U4)] uint collectionKeySize, IntPtr* fontFileEnumerator)
+        {
+            uint numberOfCharacters = collectionKeySize / sizeof(char);
```

Comment:
In general, it is much better to create a `ReadOnlySpan<char>()` over native input after doing basic validation. In this particular case, you're replicating pretty much exactly the same logic so I wouldn't necessarily do it.

`new string((char*)` terminates at the first null. If you create from a span, there is no walk of characters. If you know that embedded nulls are not a thing one will see, then it is a little more optimal to avoid the walking.